XCM Doc Restructure: Covering pallet-xcm (#4578)
w3f/polkadot-wiki: docs/learn/learn-xcm-pallet.md

provides a set of pre-defined, commonly used XCVM programs in the form of a set of extrinsics.

This pallet provides some default implementations for traits required by `XcmConfig`. The XCM
executor is also included as an associated type within the pallet's configuration.

Where the XCM format defines a set of instructions used to construct XCVM programs, `pallet-xcm`
defines a set of extrinsics that can be utilized to build XCVM programs, either to target the local
or external chains. `pallet-xcm`'s functionality is separated into three categories:

:::note

Remember, all XCMs are XCVM programs that follow
[the XCM format](https://github.com/paritytech/xcm-format). It is the job of the XCM executor is to
handle and execute these programs.

:::

1. Primitive, dispatchable functions to locally execute an XCM.
2. High-level, dispatchable functions for asset transfers.
3. Version negotiation-specific dispatchable functions.

## Primitive Extrinsics

There are two primary primitive extrinsics. These extrinsics handle sending and executing XCVM
programs as dispatchable functions within the pallet.

> 1. `execute` - This call contains direct access to the XCM executor. It is tje job of the executor

Reviewer: ```suggestion
1. `execute` - This call contains direct access to the XCM executor. It is the job of the executor
```

   to check the message and ensure that no barrier/filter will block the execution of the XCM. Once
   it is deemed valid, the message will then be _locally_ executed, therein returning the outcome as
   an event. This operation is executed on behalf of whichever account has signed the extrinsic.
   It's possible for only a partial execution to occur.
2. `send` - This call specifies where a message should be sent externally to a particular
   destination, i.e. a parachain, smart contract, or any system which is governed by consensus. As

>    before with `execute`, the Executor checks the origin, destination, and message and is then sent

Reviewer: No? I don't think the local XCM executor is consulted before the XCM is sent.

Reviewer: https://github.com/paritytech/polkadot/blob/5feae5d7e5630e9e3837a2f9517c8ae0a376dea4/xcm/pallet-xcm/src/lib.rs#LL769C5-L769C5

Author: A mistake on my part - only for local execution is the executor called. Send merely sends it _to be_ executed on the destination.

be sent, and which XCM transport protocol to use. For example, Kusama, the canary network, uses the
`ChildParachainRouter` which only allows for Downward Message Passing from the relay to parachains
to occur.

You can read more about [XCM transport methods here](./learn-xcm-transport.md).

:::

## Asset Transfer Extrinsics

Several extrinsics within the pallet handle asset transfer logic. They define a predetermined set of
instructions for sending and executing XCMs. Two variants of these functions are prefixed with
`limited_`. They have the same functionality but can specify a weight to pay for the XCM fee.

Otherwise, the fee is taken as needed from the asset being transferred.

1. `reserve_transfer_assets` - Transfer some assets from the local chain to the sovereign account of
   a destination chain and forward an XCM containing a `ReserveAssetDeposited` instruction, which
   serves as a notification.
2. `teleport_assets` - Teleport some assets from the local chain to some destination chain.

### Transfer Reserve vs. Teleport

While both extrinsics deal with transferring assets, they exhibit fundamentally different behavior.

> - **Teleporting** an asset implies a two-step process: the asset is burned/destroyed in the origin

Reviewer: They're not necessarily always burnt or destroyed; what is always true though is that the assets being teleported are removed from active circulation.

Author: I guess this is subjective to how one implements [`check_out`](https://github.com/paritytech/polkadot/blob/master/xcm/xcm-executor/src/traits/transact_asset.rs#L66), but I will add that detail.

  chain and re-minted to whatever account is specified at the destination. Teleporting should only
  occur if there is an inherent and bilateral trust between the two chains, as the tokens destroyed
  _could not_ necessarily be guaranteed to have the same properties when minted at the destination,
  as well as that there has to be **trust** that the a particular chain burned, or re-minted the
  assets.
- **Transferring** or **reserving** an asset implies that **equivalent** assets (i.e, native
  currency, like `DOT` or `KSM`) are withdrawn from _sovereign account_ of the origin chain and
  deposited into the sovereign account on the destination chain. Unlike teleporting an asset, it is
  not destroyed and re-minted, rather a trusted, 3rd entity is used (i.e., Statemint on Polkadot) to

>   **reserve** the assets, wherein the sovereign account on the destination chain obtains ownership

Reviewer: The sovereign account is located on the reserve chain, not the destination chain.

Reviewer: I think you meant to say:
```suggestion
  **reserve** the assets, wherein the sovereign account of the destination chain on the reserve chain obtains ownership
```

  of these assets.

:::info

A sovereign account refers to an account within a particular consensus system. Even though accounts
may be different in terms of factors such as address, XCM agnostic nature enables for communication
between these sovereign accounts that are in different consensus systems.

:::

## Version Negotiation Extrinsics

The following extrinsics require root, as they are only used when bypassing XCM version negotiation.
They change any relevant storage aspects that enforce anything to do with XCM version negotiations.

1. `force_xcm_version` - Modifies the `SupportedVersion` storage to change a particular
   destination's stated XCM version.
2. `force_default_xcm_version` - Modifies the `SafeXcmVersion` storage, which stores the default

>    version when the destination's version is unknown.

Reviewer: ```suggestion
   XCM version to use when the destination's version is unknown.
```

programs as dispatchable functions within the pallet.

1. `execute` - This call contains direct access to the XCM executor. It is the job of the executor
   to check the message and ensure that no barrier/filter will block the execution of the XCM. Once
   it is deemed valid, the message will then be _locally_ executed, therein returning the outcome as
   an event. This operation is executed on behalf of whichever account has signed the extrinsic.
   It's possible for only a partial execution to occur.
2. `send` - This call specifies where a message should be sent
   ([via a transport method](./learn-xcm-transport.md)) externally to a particular destination, i.e.
   a parachain, smart contract, or any system which is governed by consensus. In contrast to
   `execute`, the executor is not called locally, as the execution will occur on the destination
   chain.

:::info

The XCM pallet needs the `XcmRouter` to send XCMs. It is used to dictate where XCMs are allowed to
be sent, and which XCM transport protocol to use. For example, Kusama, the canary network, uses the
`ChildParachainRouter` which only allows for Downward Message Passing from the relay to parachains
to occur.

> You can read more about [XCM transport methods here](./learn-xcm-transport.md).

Reviewer: ```suggestion
You can read more about [XCM transport mechanisms here](./learn-xcm-transport.md).
```
Or "protocols", but not "methods" as it conflicts with the naming of class/object methods.

Author: Used "protocols", as we reference them as such in other places in the wiki.

The XCM pallet ([`pallet-xcm`](https://github.com/paritytech/polkadot/tree/master/xcm/pallet-xcm))
provides a developer-friendly interface for most of the common XCM messages.

This pallet provides some default implementations for some traits required by `XcmConfig`, as well

> as an instance of the XCM Executor provided as a trait within the pallet's own configuration.

Reviewer: If it's provided as a trait, then it's not providing an instance, because by its very nature, traits are abstract concepts and not concrete implementations.

`pallet-xcm` provides a default interface in the form of a pallet, that can manage and deal with

> XCM-related storage and higher-level dispatchable functions.

Reviewer: What is this talking about? Default interface between which two systems? And what kind of XCM-related storage?

Author: It was meant to show that the pallet is used to provide a common/default interface for developers to interact with XCM, but perhaps this wording is incorrect. 

> And what kind of XCM-related storage?

Storage related to versioning, i.e `SafeXcmVersion`.

Author: Would it be more accurate to assume that the pallet is a 'wrapper' for the `xcm-executor`?

Reviewer: It should say that it is an interface between FRAME and XCM. I would not even say that it is the default, because the transport pallets all require specifying an XCM executor.

Author: Reworded 👍

It defines a set of extrinsics that can be utilized to build XCVM programs, either to target the

> local or external chains. `pallet-xcm`'s functionality is separated into three categories:

Reviewer 2: The XCM format defines a set of _instructions_ that can be utilized to build XCVM programs. The _extrinsics_ are pre-written, commonly used programs.

Author: I think the notion that XCMs == XCVM programs need to be pushed more, as sometimes the nomenclature implies they are something akin to an arbitrary packet (rather than a set of instructions that need a VM).

:::note

> Remember, all XCM messages are effectively XCVM programs that contain a set of instructions. It is

Reviewer: ```suggestion
Remember, all XCMs are effectively XCVM programs that contain a set of instructions. It is
```

I know we say "XCM messages" colloquially, but do remember that the term is an oxymoron, because the M in XCM already stands for message.

Author: I think it would be preferable to use XCMs over XCM messages, I was debating which to use throughout.

> the job of the XCM exector to handle these programs, and their subsequent instructions as needed.

Reviewer: ```suggestion
the job of the XCM executor to handle these programs.
```

I'd probably delete that entire sentence because it's unclear what the subsequent instructions are.

:::

1. Primitive, dispatchable functions to locally execute an XCM message.
2. High-level, dispatchable functions for asset transfers.
3. Version negotiation-specific dispatchable functions.

## Primitive Extrinsics

> There are two primary primitive extrinsics. These extrinsics handle sending and executing XCVM

Reviewer: Is "primary primitive extrinsic" a technical term? Who designated `execute` and `send` as "primary primitive extrinsics"?

Author: Within the [v2 messaging audit](https://blog.quarkslab.com/resources/2022-02-27-xcmv2-audit/21-12-908-REP.pdf), they were called 'primitive' functions.

programs as dispatchable functions within the pallet.

1. `execute` - This call is direct access to the XCM executor. It checks the origin and message and

>    ensures that no barrier/filter will block the execution of the XCM message. Once it is deemed

Reviewer: I don't think the `execute` extrinsic does this -- the XCM executor does.

Author: yes, I meant for it to be executor - will adjust to be clearer. The `execute` extrinsic merely calls the executor (`T::XcmExecutor::execute_xcm_in_credit`).

   valid, the message will then be _locally_ executed, therein returning the outcome as an event.
   This operation is executed on behalf of whichever account has signed the extrinsic. It's possible
   for only a partial execution to occur.
2. `send` - This call specifies where a message should be sent externally to a particular

>    destination, i.e., another parachain. It checks the origin, destination, and message and is then

Reviewer 2: Not necessarily another parachain. It could also be a smart contract on the local chain, or another consensus system entirely.

   sent to the `XcmRouter`.

:::info

The XCM pallet needs the `XcmRouter` to send messages. It is used to dictate where messages are
allowed to be sent, and which XCM transport protocol to use. For example, Kusama, the canary

> network, uses the `ChildParachainRoute` which only allows for Downward Message Passing from the

Reviewer: ```suggestion
network, uses the `ChildParachainRouter` which only allows for Downward Message Passing from the
```

relay to parachains to occur.

> You can read more about [XCM transport methods here.](./learn-xcm-transport.md)

Reviewer 2: ```suggestion
You can read more about [XCM transport methods here](./learn-xcm-transport.md).
```

:::

## Asset Transfer Extrinsics

Several extrinsics within the pallet handle asset transfer logic. They define a predetermined set of
instructions for sending and executing XCM messages. Two variants of these functions are prefixed
with `limited_`. They have the same functionality but can specify a weight to pay for the XCM fee.

Otherwise, the fee is taken as needed from the asset being transferred.

1. `reserve_transfer_assets` - Transfer some assets from the local chain to the sovereign account of

>    a destination chain and forward a notification XCM.

Reviewer: ```suggestion
   a destination chain and forward an XCM containing a `ReserveAssetDeposited` instruction.
```

2. `teleport_assets` - Teleport some assets from the local chain to some destination chain.

> ### Transfer Reserve vs Teleport

Reviewer 2: ```suggestion
### Transfer Reserve vs. Teleport
```

While both extrinsics deal with transferring assets, they exhibit fundamentally different behavior.

- **Teleporting** an asset implies a two-step process: the asset is burned/destroyed in the origin
  chain and re-minted to whatever account is specified at the destination. Teleporting should only
  occur if there is an inherent and bilateral trust between the two chains, as the tokens destroyed

>   _could not_ necessarily be guaranteed to have the same properties when minted at the destination.

Reviewer: This isn't the fundamental problem -- the more fundamental problem is whether or not the two chains can trust each other to have burnt or minted the funds beforehand. A chain can simply send the `ReceiveTeleportedAsset` XCM instruction without previously having burnt any.

- **Transferring** or **reserving** an asset implies the movement of funds from one authority to
  another, in this case, from the Origin of the sender's chain and into the sovereign account (who

>   will ultimately benefit from the assets) on the destination chain.

Reviewer 2: This is not really clear

Author: > - **Transferring** or **reserving** an asset implies that **equivalent** assets (i.e, native currency) are withdrawn from the ownership of the current origin and deposited into the sovereign account on the destination chain.  Unlike teleporting an asset, it is not destroyed, rather a trusted entity is used (i.e., Statemint on Polkadot) to **reserve** the assets, wherein the sovereign account on the destination chain obtains ownership of them.

Is this clearer? Might be worth creating a diagram for to show the difference.

Reviewer: Is the term equivalent asset defined anywhere else other than here?

Author: Under the `xcm-format` repo, it's used a few times: https://github.com/paritytech/xcm-format#transferreserveasset

Is it worth noting that this asset is representative of the origin chain? i.e, DOT becoming known as "wrapped" DOT on the destination chain, or is that inaccurate?

## Version Negotiation Extrinsics

The following extrinsics require root, as they are only used when bypassing XCM version negotiation.
They change any relevant storage aspects that enforce anything to do with XCM version negotiations.

1. `force_xcm_version` - Modifies the `SupportedVersion` storage to change a particular
   destination's stated XCM version.
2. `force_default_xcm_version` - Modifies the `SafeXcmVersion` storage, which stores the default
   version when the destination's version is unknown.
3. `force_subscribe_version_notify` - Sends an XCM message with a `SubscribeVersion` instruction to
   a destination.
4. `force_unsubscribe_version_notify` - Sends an XCM message with a `UnsubscribeVersion` instruction
   to a destination.

## Fees in the XCM Pallet

Message fees are only paid if the interior location is **not** equal to the interpreting consensus
system (known as `Here` in the context of an XCM Multilocation). Otherwise, the chain bears the

> fees. If applicable, fees are withdrawn from the assets from the specified `MultiLocation`.

Reviewer 2: Not really, there are a lot of barrier conditions for when to allow unpaid execution, and there are other ways to pull fees into the holding register, like claiming a trap.

Author: I was planning to cover this in another part of the Wiki, however I'll add some context around those scenarios here as well.
